api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirAllergyIntoleranceDaoImpl.java

```diff
@@ -10,28 +10,181 @@
 import org.openmrs.module.fhir2.api.dao.FhirAllergyIntoleranceDao;
 import org.springframework.stereotype.Component;
 
 @Component
 @Setter(AccessLevel.PACKAGE)
-public class FhirAllergyIntoleranceDaoImpl implements FhirAllergyIntoleranceDao {
+public class FhirAllergyIntoleranceDaoImpl extends BaseDaoImpl implements FhirAllergyIntoleranceDao {
 	
 	@Inject
 	@Named("sessionFactory")
 	private SessionFactory sessionFactory;
 	
+	@Inject
+	private FhirConceptService conceptService;
+	
+	@Inject
+	private FhirGlobalPropertyService globalPropertyService;
+	
 	@Override
 	public Allergy getAllergyIntoleranceByUuid(String uuid) {
 		return (Allergy) sessionFactory.getCurrentSession().createCriteria(Allergy.class).add(eq("uuid", uuid))
 		        .uniqueResult();
 	}
+	
+	@Override
+	public Collection<Allergy> searchForAllergies(ReferenceParam patientReference, StringOrListParam category,
+	        TokenOrListParam allergen, TokenOrListParam severity, TokenOrListParam manifestationCode,
+	        StringOrListParam clinicalStatus) {
+		Criteria criteria = sessionFactory.getCurrentSession().createCriteria(Allergy.class);
+		handlePatientReference(criteria, patientReference, "patient");
+		handleAllergenCategory("allergen.allergenType", category);
+		handleAllergen(criteria, allergen);
+		handleSeverity(criteria, severity).ifPresent(criteria::add);
+		handleManifestation(criteria, manifestationCode);
+		handleBoolean("voided", createClinicalStatusToken(clinicalStatus));
+		
+		return criteria.list();
+	}
+	
+	private void handleManifestation(Criteria criteria, TokenOrListParam manifestation) {
+		if (manifestation == null) {
+			return;
+		}
+		criteria.createAlias("reactions", "r");
+		criteria.createAlias("r.reaction", "c");
+		
+		handleOrListParamBySystem(manifestation, (system, tokens) -> {
+			if (system.isEmpty()) {
+				return Optional.of(in("c.uuid", tokensToList(tokens)));
+			} else {
+				if (!containsAlias(criteria, "rs")) {
+					criteria.createAlias("c.uuid", "rs");
+				}
+				
+				return Optional.of(or(eq("rs.name", system), in("c.uuid", tokensToList(tokens))));
+			}
+		}).ifPresent(criteria::add);
+	}
+	
+	private void handleAllergen(Criteria criteria, TokenOrListParam allergen) {
+		if (allergen == null) {
+			return;
+		}
+		
+		criteria.createAlias("allergen.codedAllergen", "c");
+		
+		handleOrListParamBySystem(allergen, (system, tokens) -> {
+			if (system.isEmpty()) {
+				return Optional.of(in("c.uuid", tokensToList(tokens)));
+			} else {
+				if (!containsAlias(criteria, "cs")) {
+					criteria.createAlias("c.uuid", "cs");
+				}
+				
+				return Optional.of(or(eq("cs.name", system), in("c.uuid", tokensToList(tokens))));
+			}
+		}).ifPresent(criteria::add);
+	}
+	
+	private Optional<Criterion> handleSeverity(Criteria criteria, TokenOrListParam severityParam) {
+		if (severityParam == null) {
+			return Optional.empty();
+		}
+		criteria.createAlias("severity", "c");
+		
+		return handleOrListParam(severityParam, token -> {
```

Comment:
This is a proper way to implement this; however, it is probably preferable to cache the values of the calls to `getConceptId(globalPropertyService.getGlobalProperty(...))` outside of the lambda expression. something like this:

```java
final int mildConceptId = getConceptId(globalPropertyService.getGlobalProperty(FhirConstants.GLOBAL_PROPERTY_MILD));
...
return handleOrListParam(severityParam, token -> {
    try {
        AllergyIntolerance.AllergyIntoleranceSeverity severity = AllergyIntolerance.AllergyIntoleranceSeverity
            .fromCode(token.getValue());
        
        switch(servity) {
             case MILD:
                 return Optional.of(eq("c.conceptId", mildConceptId));
             ...
        }
    }
    ...
});
```

Thinking about this some more, we should probably create a method on the `globalPropertyService` called `getGlobalProperties(String... properties)` that looks something like this:

```java
public Collection<String> getGlobalProperties(String... properties) {
    List<GlobalProperty> globalProperties = (GlobalProperty) sessionFactory.getCurrentSession()
        .createCriteria(GlobalProperty.class).add(Restrictions.in("property", properties)).list();
    return globalProperties.stream().filter(p -> p != null).map(GlobalProperty::getPropertyValue).collect(Collectors.toList());
}
```

Then we can use this:

```java
final int[] severityConceptUuids = globalPropertyService.getGlobalProperties(FhirConstants.GLOBAL_PROPERTY_MILD, FhirConstants.GLOBAL_PROPERTY_MODERATE, FhirConstants.GLOBAL_PROPERTY_SEVERE, FhirConstants.GLOBAL_PROPERTY_OTHER);
...
return handleOrListParam(severityParam, token -> {
    try {
        AllergyIntolerance.AllergyIntoleranceSeverity severity = AllergyIntolerance.AllergyIntoleranceSeverity
            .fromCode(token.getValue());
        
        switch(servity) {
             case MILD:
                 return Optional.of(eq("c.uuid", severityConceptIds[0]));
             ...
        }
    }
    ...
});
```